File: go/SampleSender.py
"""A simple script to send some synthetic values to the MeterReceiver server to test it."""

import MeterReader_pb2
import random
import socket
import struct
import time
import binascii

TCP_IP = '127.0.0.1'
TCP_PORT = 2110
MAGIC_STRING = 'AA'

def send_message(sock, message):
    """ Send a serialized message (protobuf Message interface)
        to a socket in the following format:
          * the magic header
          * the length of the encoded protobuf message in 4 bytes (big endian)
          * the encoded protobuf message
          * the crc32 value of the protobuf message in 4 bytes (big endian)
    """
    s = message.SerializeToString()
    packed_len = struct.pack('>L', len(s))

    crc = struct.pack('>L', (binascii.crc32(s) & 0xFFFFFFFF))

    sock.sendall(MAGIC_STRING + packed_len + s + crc)

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((TCP_IP, TCP_PORT))
print "Connected"

def send_garbage(sock, amount):
    for i in range(amount):
        sock.send(chr(random.randint(0, 255)))


series=0
count=0
while True:
    msg = MeterReader_pb2.Message()

    msg.update.meterId = 1
    if random.randint(0,10) == 1:
        # new series
        series += 1
        count = 0

    count += random.randint(0, 10)

    msg.update.seriesId = series
    msg.update.currentCounterValue = count

    send_message(s, msg)
    if random.randint(0, 10) < 3:
        send_garbage(s, random.randint(1, 50))

    time.sleep(1)



s.close()

File: go/SampleSender.go
package main

import (
	"MeterReader"
	"io"
	"log"
	"math/rand"
	"net"
	"time"
)

const host string = "localhost:2110"

// Setup
var NUM_METERS              int = 4
var UPDATE_INTERVAL_SECONDS int = 5
var NUM_GARBAGE_BYTES       int = 49

//----------------------------------------------------------------------
func meterId() *uint32 {
  var x uint32 = uint32(1+rand.Intn(NUM_METERS))
  return &x
}

func send_garbage(sock io.Writer, amount int) {
	var garbage []byte
	garbage = make([]byte, amount, amount)

	for i := 0; i < amount; i++ {
		garbage[i] = byte(rand.Intn(255))
	}

	sock.Write(garbage)
}

func main() {
	s, err := net.Dial("tcp", host)
	if err != nil {
		log.Fatalf("ERROR: Unable to connect to %s", host)
	}

	var series uint32
	var count uint64
	for {
		update := new(MeterReader.CounterUpdate)
		update.MeterId = meterId()
		if rand.Intn(10) == 1 {
			// new series
			series += 1
			count = 0
		}

		count += uint64(rand.Int63n(10))

		update.SeriesId = &series
		update.CurrentCounterValue = &count

		MeterReader.SendUpdate(s, update)
		if rand.Intn(10) < 3 {
			send_garbage(s, rand.Intn(NUM_GARBAGE_BYTES)+1)
		}

		time.Sleep(time.Duration(rand.Intn(UPDATE_INTERVAL_SECONDS)) * time.Second)
	}
}
